### PR TITLE
PowerShell 6, 7 images and pipeline and Documentation

### PR DESCRIPTION
**Pull request by TsuyoshiUshio**
This PR is follow up PR of this PR. After the PR has been published, a lot of change made to master, to avoid rollback, I create a new PR from scratch, however, this PR is based on the PR, and expand to PowerShell7. 

https://github.com/Azure/azure-functions-docker/pull/253

I implement all request from @AnatoliB 

Could you have a look? 

* PowerShell6 Core Tools (Azure CLI, AzModule, Core Tools V3)
* PowerShell7 Core Tools (Azure CLI, AzModule, Core Tools V3)
* Pipeline Build, Test, Publish
* Documentation

The Pipeline is already tested.  The pipeline will run 10:30. 

image

### CODE REVIEW

File: host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+ADD ${PS_PACKAGE_URL} /tmp/powershell.deb
+
+# Define ENVs for Localization/Globalization
+ENV DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=false \
+    LC_ALL=en_US.UTF-8 \
+    LANG=en_US.UTF-8 \
+    # set a fixed location for the Module analysis cache
+    PSModuleAnalysisCachePath=/var/cache/microsoft/powershell/PSModuleAnalysisCache/ModuleAnalysisCache
+
+ENV FUNCTIONS_WORKER_RUNTIME=powershell
```
Comment (AnatoliB) [inline host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile:28]: Please also set FUNCTIONS_WORKER_RUNTIME_VERSION to ~6.

File: host/3.0/buster/amd64/powershell/powershell7-core-tools.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ADD ${PS_PACKAGE_URL} /tmp/powershell.deb
+
+# Define ENVs for Localization/Globalization
+ENV DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=false \
+    LC_ALL=en_US.UTF-8 \
+    LANG=en_US.UTF-8 \
+    # set a fixed location for the Module analysis cache
+    PSModuleAnalysisCachePath=/var/cache/microsoft/powershell/PSModuleAnalysisCache/ModuleAnalysisCache
+
+ENV FUNCTIONS_WORKER_RUNTIME=powershell
```
Comment (AnatoliB) [inline host/3.0/buster/amd64/powershell/powershell7-core-tools.Dockerfile:28]: Please also set FUNCTIONS_WORKER_RUNTIME_VERSION to ~7.

File: host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        }" \
+    # install AzModule
+    && pwsh \
+        -NoLogo \
+        -NoProfile \
+        -Command " \
+        Set-PSRepository PSGallery -InstallationPolicy Trusted; \
+        Install-Module -Name Az -AllowClobber -Scope AllUsers -Confirm:\$False -Force" \
+    # initialize powershell module cache
+    && pwsh -NoLogo -NoProfile -Command "try { IAmSureThisCommandDoesNotExist } catch { exit 0 }" \
```
Comment (AnatoliB) [inline host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile:95]: I'm wondering why `exit 0`. Why not just ignore the exception and do nothing?

Comment (AnatoliB) [inline host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile:95]: After running the non-existing command, we need to wait for the cache file to be flushed to the file, like you do in lines 76-86. Also, there is no reason to wait for the cache update before installing the Az module, so I suggest moving the waiting logic here.

File: host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        -NoLogo \
+        -NoProfile \
+        -Command " \
+        \$ErrorActionPreference = 'Stop' ; \
+        \$ProgressPreference = 'SilentlyContinue' ; \
+        while(!(Test-Path -Path \$env:PSModuleAnalysisCachePath)) {  \
+            Write-Host "'Waiting for $env:PSModuleAnalysisCachePath'" ; \
+            Start-Sleep -Seconds 6 ; \
+        }" \
+    # install AzModule
```
Comment (AnatoliB) [inline host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile:87]: Cosmetics: "Az module"

File: host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile (hunk 3 above)
Comment (AnatoliB) [inline host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile:95]: Could you please add a comment explaining why we are running this strange command, something like: "Invoke a non-existing command to force PowerShell to perform complete module analysis."

File: host/3.0/core-tools-publish.yml [hunk 5]
```diff
@@ -464,3 +519,69 @@ stages:
...
+          ActualVersion=`docker run $TARGET_REGISTRY/powershell:$(MajorVersion)-powershell6-core-tools func --version`
+          if [[ "$ActualVersion" != "$(TargetVersion)" ]]; then
+            echo "unmatch the ActualVersion and TargetVersion ${ActualVersion} and $(TargetVersion)";
+            exit 1;
+          fi
+          docker run $TARGET_REGISTRY/powershell:$(MajorVersion)-powershell6-core-tools az --version
+
+          AZ_MODULE_COUNT=`docker run $TARGET_REGISTRY/powershell:$(MajorVersion)-powershell6-core-tools pwsh  -Command "(Get-Module -ListAvailable Az*).count"`
+          if [ ! "$AZ_MODULE_COUNT" -gt "0" ]; then
+            echo "can not find Az Module for powershell6-core-tools. The number of the AzModule is $AZ_MODULE_COUNT"
```
Comment (AnatoliB) [inline host/3.0/core-tools-publish.yml:559]: Cosmetics: "Az module"

File: host/3.0/core-tools-publish.yml [hunk 6]
```diff
@@ -464,3 +519,69 @@ stages:
...
+            exit 1;
+          fi
+          ActualVersion=`docker run $TARGET_REGISTRY/powershell:$(MajorVersion)-powershell6-core-tools func --version`
+          if [[ "$ActualVersion" != "$(TargetVersion)" ]]; then
+            echo "unmatch the ActualVersion and TargetVersion ${ActualVersion} and $(TargetVersion)";
+            exit 1;
+          fi
+          docker run $TARGET_REGISTRY/powershell:$(MajorVersion)-powershell6-core-tools az --version
+
+          AZ_MODULE_COUNT=`docker run $TARGET_REGISTRY/powershell:$(MajorVersion)-powershell6-core-tools pwsh  -Command "(Get-Module -ListAvailable Az*).count"`
```
Comment (AnatoliB) [inline host/3.0/core-tools-publish.yml:557]: Querying for `Az*` will find modules with names like `AzTable`, `AzureRM`, or `AzSomethingElse` if they happen to be accidentally installed. You want to check for `Az` here, so let's be specific.

File: host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile (hunk 1 above)
Comment (TsuyoshiUshio) [reply host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile:28]: Done

File: host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile (hunk 3 above)
Comment (TsuyoshiUshio) [reply host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile:96]: Done

Comment (TsuyoshiUshio) [reply host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile:96]: @AnatoliB if we pass it, it return status code 1. so it fails pipeline. 

```
$ pwsh -NoLogo -NoProfile -Command "try { IAmSureThisCommandDoesNotExist } catch {  }"
$ echo $?
1
$ pwsh -NoLogo -NoProfile -Command "try { IAmSureThisCommandDoesNotExist } catch { exit 0 }"
$ echo $?
0
```

Comment (TsuyoshiUshio) [reply host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile]: Done

File: host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    && apt-get install -y azure-cli azure-functions-core-tools-3 \
+    #
+    # Create a non-root user to use if preferred - see https://aka.ms/vscode-remote/containers/non-root-user.
+    && groupadd --gid $USER_GID $USERNAME \
+    && useradd -s /bin/bash --uid $USER_UID --gid $USER_GID -m $USERNAME \
+    # [Optional] Add sudo support for the non-root user
+    && apt-get install -y sudo \
+    && echo $USERNAME ALL=\(root\) NOPASSWD:ALL > /etc/sudoers.d/$USERNAME\
+    && chmod 0440 /etc/sudoers.d/$USERNAME \
+    # intialize powershell module cache
```
[the remark this answers is not in the document]
Comment (TsuyoshiUshio) [reply host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile]: Done

File: host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile (hunk 4 above)
Comment (TsuyoshiUshio) [reply host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile]: Done

File: host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile (hunk 3 above)
Comment (TsuyoshiUshio) [reply host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile]: Done

File: host/3.0/core-tools-publish.yml (hunk 5 above)
Comment (TsuyoshiUshio) [reply host/3.0/core-tools-publish.yml:559]: Done

File: host/3.0/core-tools-publish.yml (hunk 6 above)
Comment (TsuyoshiUshio) [reply host/3.0/core-tools-publish.yml]: Done

**Comment by TsuyoshiUshio** [comment]
Hi @AnatoliB 
I follow your advice. Only one answer left.   I tested the pipeline works. 
Could you have a look?

File: host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile (hunk 3 above)
Comment (AnatoliB) [reply host/3.0/buster/amd64/powershell/powershell6-core-tools.Dockerfile]: I see, you are right, thank you.

**Comment by TsuyoshiUshio** [comment]
I've got the review of Anatoli, Hi @ahmelsayed or @Hazhzeng , could you have a look and merge this?

File: host/3.0/core-tools-publish.yml [hunk 8]
```diff
@@ -464,3 +519,69 @@ stages:
...
+          ActualVersion=`docker run $TARGET_REGISTRY/powershell:$(MajorVersion)-powershell6-core-tools func --version`
+          if [[ "$ActualVersion" != "$(TargetVersion)" ]]; then
+            echo "unmatch the ActualVersion and TargetVersion ${ActualVersion} and $(TargetVersion)";
+            exit 1;
+          fi
+          docker run $TARGET_REGISTRY/powershell:$(MajorVersion)-powershell6-core-tools az --version
+
+          AZ_MODULE_COUNT=`docker run $TARGET_REGISTRY/powershell:$(MajorVersion)-powershell6-core-tools pwsh  -Command "(Get-Module -ListAvailable Az).count"`
+          if [ ! "$AZ_MODULE_COUNT" -gt "0" ]; then
+            echo "can not find Az Module for powershell6-core-tools. The number of the AzModule is $AZ_MODULE_COUNT"
```
Comment (AnatoliB) [inline host/3.0/core-tools-publish.yml:559]: Cosmetics: "Az module" ("module" is just a common noun here, no need to capitalize it or merge with "Az").

Comment (TsuyoshiUshio) [reply host/3.0/core-tools-publish.yml:559]: Good catch! I'd like to fix it now.

Comment (TsuyoshiUshio) [reply host/3.0/core-tools-publish.yml]: Done!

**Pull request merged**